packages/base/src/state/bond_providers.rs

```diff
@@ -0,0 +1,232 @@
+use crate::error::core::{ContractError, ContractResult};
+use cosmwasm_std::{Addr, Empty, StdResult, Storage};
+use cw_storage_plus::{Item, Map};
+
+pub struct BondProviders<'a> {
+    pub providers: Map<'a, Addr, Empty>,
+    pub next_provider_ptr: Item<'a, u64>,
+}
+
+impl<'a> BondProviders<'a> {
+    pub const fn new(storage_key: &'a str) -> Self {
+        Self {
+            providers: Map::new(storage_key),
+            next_provider_ptr: Item::new("bond_providers_ptr"),
+        }
+    }
+
+    pub fn init(&self, storage: &mut dyn Storage) -> StdResult<()> {
+        self.next_provider_ptr.save(storage, &0)?;
+        Ok(())
+    }
+
+    pub fn get_all_providers(&self, storage: &dyn Storage) -> StdResult<Vec<Addr>> {
+        let providers = self
+            .providers
+            .range(storage, None, None, cosmwasm_std::Order::Ascending)
+            .collect::<StdResult<Vec<(Addr, Empty)>>>()?
+            .into_iter()
+            .map(|(addr, _)| addr)
+            .collect();
```

Comment:
```suggestion
        let providers = self
            .providers
            .range(storage, None, None, cosmwasm_std::Order::Ascending)
            .map(|x| x.map(|(addr, _)| addr))
            .collect::<StdResult<Vec<_>>>()?;
```

Avoid creating extra vector and draining it